src/Nerdbank.MessagePack/SecureHash/ByValueEnumerableEqualityComparer`2.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Andrew Arnott. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+
+namespace Nerdbank.MessagePack.SecureHash;
+
+/// <summary>
+/// An equality comparer that performs deep by-value comparisons.
+/// </summary>
+/// <typeparam name="TEnumerable">The type of the enumerable.</typeparam>
+/// <typeparam name="TElement">The type of element in the enumerable.</typeparam>
+/// <param name="equalityComparer">The equality comparer for individual elements.</param>
+/// <param name="getEnumerable">The function that gets the enumerable of the collection.</param>
+internal class ByValueEnumerableEqualityComparer<TEnumerable, TElement>(
+	IEqualityComparer<TElement> equalityComparer,
+	Func<TEnumerable, IEnumerable<TElement>> getEnumerable) : IEqualityComparer<TEnumerable>
+{
+	/// <inheritdoc/>
+	public bool Equals(TEnumerable? x, TEnumerable? y)
+	{
+		if (x is null || y is null)
+		{
+			return ReferenceEquals(x, y);
+		}
+
+		IEnumerable<TElement> enumerableX = getEnumerable(x);
```

Comment:
Specializing a comparer for types implementing `IList<T>` should eliminate enumerator allocations for a lot of types.

Reply:
#65